typing/typecore.ml

```diff
@@ -65,6 +65,24 @@ type wrong_name = {
   valid_names: string list;
 }
 
+type wrong_kind_context =
+  | Pattern
+  | Expression of type_forcing_context option
+
+type wrong_kind_sort =
+  | Constructor
+  | Record
+  | Boolean
+  | List
+  | Unit
+
+let wrong_kind_sort_of_constructor (lid : Longident.t) =
+  match lid with
+  | Lident "true" | Lident "false" | Ldot(_, "true") | Ldot(_, "false") -> Boolean
+  | Lident "[]" | Lident "::" | Ldot(_, "[]") | Ldot(_, "::") -> List
+  | Lident "()" | Ldot(_, "()") -> Unit
+  | _ -> Constructor
```

Comment:
I am a bit troubled by the purely syntactic logic here. It is arguably not wrong to think that a user-described datatype that was defined to use `::` as constructor can be referred to as a list, but it is also not exactly correct. I think in similar situations we have argued that the compiler should not embed specialized knowledge of this kind, even for error messages. (But maybe it was for arithmetic operators that are "more library" than those names?)

Reply:
Another weak argument against this specialization is that alternative syntaxes are not required to have `list`, `unit` or `boolean` literals.

Reply:
I very deliberately use the terms "list literal", "boolean literal" and "unit literal" in these error messages. These are intended to describe the piece of syntax, not the type that the syntax may result in, and I've tried to stick close to the terminology used for similar constructs in the manual.

Without this specialisation, the messages essentially leak the fact that e.g. `()` is a variant constructor into an error message that is likely seen by beginners. Whereas, an advanced user who knows that you can make a specialised list type, will not be confused by the term "list literal" for e.g. `[a; b; c]`.

Reply:
I find your argument convincing. The point of @Octachron remains that we are embedding knowledge about the surface syntax in the type-checker, which is unfortunate. A correct long-term fix for this issue would probably be to create explicit parsetree node for those constructs, instead of desugaring in the parsing; but this is orthogonal to the present change.

Is there a simple change that would alleviate this issue? For example, possibly this "literal unparsing" could be moved to a helper function in a `parsing/` file. (I see that `parsing/ast_mapper.ml` already has logic to this effect, that could be shared among several consumers. But then it doesn't use the `Ldot` generalization, so code reuse seems difficult.) I don't see a simple suggestion to make, so I would be fine with leaving things as-is, but maybe @lpw25 or @Octachron have better ideas.